lib/embulk/output/bigquery/value_converter_factory.rb

```diff
@@ -224,6 +224,13 @@ def string_converter
                 val # Users must care of BQ timestamp format
               }
             end
+          when 'TIME'
+            Proc.new {|val|
+              next nil if val.nil?
+              with_typecast_error(val) do |val|
+                Time.parse(val).strftime("%H:%M:%S.%6N")
+              end
```

Comment:
I think the TIME type doesn't consider the offset and simply processes the time portion. So this code converts all ```2024-07-24 10:00:00 +09:00```,  ```2024-07-24 10:00:00 -09:00```, ```10:00:00``` to ```10:00:00```.

Your comment suggests that ```2024-07-24 10:00:00 +09:00``` should be converted to ```19:00:00```, and ```2024-07-24 10:00:00 -09:00``` should be converted to ```01:00:00```. And even if the timezone is ```+09:00```, the input value ```10:00:00``` should be output as ```10:00:00``` without any changes.

Is this understanding correct?

Reply:
For example, a user has a timestamp data like '2024-07-25 9:00:00 +9:00'.
For some reason, the user wants to convert the DATE and TIME fields in UTC (and also insert the same data into the DATETIME field).
In this case, data will be:
* DATE `2024-07-25`(UTC)
* DATETIME `2024-07-25 00:00:00Z`
* TIME: `00:00:00`(UTC)

If a user wants to insert in the `+9:00` timezone,
* DATE `2024-07-25`(+09:00)
* DATETIME `2024-07-25 09:00:00` (+09:00)
* TIME: `15:00:00` (+09:00)

IIUC, DATA and DATETIME fields care about that. so I think It is necessary to care about the `TIME` fields too.

Does this PR work like this? If, so I'll check like this PR.

[Attached codes](https://github.com/embulk/embulk-output-bigquery/pull/168/files#r1689748990) was just for example, It will be `TimeWithZone.set_zone_offset(Time.strptime(val, @timestamp_format), zone_offset).strftime("%H:%M:%S.%6N")`

Reply:
i fixed, but test case that was previously validated successfully has now failed..
Could you check the code?

Reply:
I will check the code. the reason for the error is you set the timestamp format explicitly. This code expects the input data format `%H:%M:%S`.

```ruby
          converter = ValueConverterFactory.new(
            SCHEMA_TYPE, 'TIME',
            timestamp_format: '%H:%M:%S'
          ).create_converter
```

Reply:
Thanks, I removed the type constraint from the test cases input. 
And I added support for the Time type in the timestamp converter after I got your example of converting '2024-07-25 9:00:00 +9:00'. 😃

Reply:
Thanks! Please wait a while, I need to understand the timestamp behavior of this plugin.
I don't understand why `TIMESTAMP` uses `TimeWithZone.set_zone_offset` and `DATETIME` doesn't use it yet.

https://github.com/embulk/embulk-output-bigquery/blob/8e7df4a8ea638a8cc5a77e991e967d40e5c6e2a4/lib/embulk/output/bigquery/value_converter_factory.rb#L192-L226

Reply:
I agree with your suggestion that ```DATETIME``` should use ```TimeWithZone.set_zone_offset```. For the same reason, I think the ```TIME``` code might also be changed in the same way.

```
          when 'DATETIME'
            if @timestamp_format
              Proc.new {|val|
                next nil if val.nil?
                with_typecast_error(val) do |val|
                  TimeWithZone.set_zone_offset(Time.parse(val), zone_offset).strftime("%Y-%m-%d %H:%M:%S.%6N")
                end
              }
            else
              Proc.new {|val|
                next nil if val.nil?
                val # Users must care of BQ timestamp format
              }
            end
          when 'TIME'
             if @timestamp_format
              Proc.new {|val|
                next nil if val.nil?
                with_typecast_error(val) do |val|
                  TimeWithZone.set_zone_offset(Time.parse(val), zone_offset).strftime("%H:%M:%S.%6N")
                end
              }
            else
              Proc.new {|val|
                next nil if val.nil?
                val # Users must care of BQ timestamp format
              }
            end
```

Reply:
@p-eye Thank you for waiting. It took a long time.

I misunderstood the `TimeWithZone` feature
https://sonots.github.io/time_with_zone/TimeWithZone.html
IIUC, `TimeWithZone`.zone_offset only changes `timezone` field.

I'm not 100% sure, but I think It is not meaningful to use `TimeWithZone` in the time part.
So, We can choose (The result will be the same whether it is there or not.)

1. Do not use `TimeWithZoe.zone_offset` and add the comment `TimeWithZone doesn't affect any change`. So, it is not used in this part.
2. Add `TimeWithZoe.zone_offset` for consistency. and add  `TimeWithZone doesn't affect any change`. So, Use consistency with other parts.

My opinion is No.2.

```
[25] pry(main)> TimeWithZone.set_zone_offset(Time.strptime("15:00:01", "%H:%M:%S"), TimeWithZone.zone_offset("Asia/Tokyo")).strftime("%H:%M:%S.%6N")
=> "15:00:01.000000"
[26] pry(main)> TimeWithZone.set_zone_offset(Time.strptime("15:00:01", "%H:%M:%S"), TimeWithZone.zone_offset("UTC")).strftime("%H:%M:%S.%6N")
=> "15:00:01.000000"
```

`2024-08-10 15:00:01 +0900` is "+9:00" offset.
It is `2024-08-10 14:00:01 +0800` in `Asia/Thaipei` timezone.
But `TimeWithZone` returns `2024-08-10 15:00:01.000000 +08:00`

```
[27] pry(main)> TimeWithZone.set_zone_offset(Time.strptime("2024-08-10 15:00:01 +0000", "%Y-%m-%d %H:%M:%S %z"), TimeWithZone.zone_offset("UTC")).strftime("%Y-%m-%d %H:%M:%S.%6N %:z")
=> "2024-08-10 15:00:01.000000 +00:00"
[29] pry(main)> TimeWithZone.set_zone_offset(Time.strptime("2024-08-10 15:00:01 +0900", "%Y-%m-%d %H:%M:%S %z"), TimeWithZone.zone_offset("Asia/Taipei")).strftime("%Y-%m-%d %H:%M:%S.%6N %:z")
=> "2024-08-10 15:00:01.000000 +08:00"
```

What do you think?

Reply:
yes, i agree with the No.2 idea.

Reply:
@p-eye  Thanks. Please add `TimeWithZoe.zone_offset` (and comment) in `TIME`. And, keep `DATETIME` as is. (I would better to add `TimeWithZoe.zone_offset`. But this code deploys for a long time I don't want to care about unexpected behavior.)

Reply:
i fixed. please check https://github.com/embulk/embulk-output-bigquery/pull/168/commits/489b6b2a1650878983db77f3bdab46e411e0b227. thanks